### ISSUE TITLE
[Question] How can I get out of bag error from lightbgm

### ISSUE THREAD

**Issue by alexhallam** [microsoft/LightGBM#5556]
This is not a bug just a question.

I just wanted to know if there was a way to get something like an 'out of bag error' from lightbgm.

I know for random forests I can easily get an out of bag error estimate. This comes from the error on the validation set that random forest is using when bagging (bootstrap aggregating).

**Comment by jameslamb** [comment]
Thanks for using LightGBM!

Random Forest implementations randomly sample rows of the training data in each tree, and so can calculate an "out of bag" error by predicting on each row using only an ensemble of the trees for which each sample wasn't chosen.

LightGBM, by default, does not work that way. With LightGBM's default settings, every row in the input data is used in every tree.

To get something like what you are describing, hold out a chunk of labeled data prior to training, and pass it to LightGBM as a validation set.

Here's an example using the R package.

```r
library(lightgbm)

data(agaricus.train, package = "lightgbm")
data(agaricus.test, package = "lightgbm")

dtrain <- lgb.Dataset(
    data = agaricus.train$data
    , label = agaricus.train$label
)

dtest <- lgb.Dataset.create.valid(
    dataset = dtrain
    , data = agaricus.test$data
    , label = agaricus.test$label
)

model <- lgb.train(
    params =  list(
        objective = "regression"
        , metric = "l2"
        , num_leaves = 7L
    )
    , data = dtrain
    , nrounds = 5L
    , valids = list(validation_set = dtest)
)
```

Produces logs like

```text
[1] "[1]:  validation_set's l2:0.203918"
[1] "[2]:  validation_set's l2:0.167247"
[1] "[3]:  validation_set's l2:0.137026"
[1] "[4]:  validation_set's l2:0.112671"
[1] "[5]:  validation_set's l2:0.0916572"
```

Hope that helps. Please let us know if you have other questions.
